<div style="background-color: #c34a4a; color: #FFFFFF; padding: 5px; text-align: center; border-radius: 15px; box-shadow: 0 4px 8px rgba(0, 0, 0, 0.5);">
  <h1 style="margin: 0; font-family: 'Helvetica Neue', Helvetica, Arial, sans-serif; font-weight: bold; font-size: 32px;"> GENDER RECOGINITION FROM FACIAL IMAGES</h1>
  <p style="margin: 10px 0; font-size: 32px; font-style: italic; text-align: center;">using Convolutional Neural Networks</p>
</div>




<div style="background-color: #2D2D2D; color: #FFFFFF; padding: 10px; text-align: center; border-radius: 15px; box-shadow: 0 4px 8px rgba(0, 0, 0, 0);">
  <h2 style="margin: 0; font-family: 'Helvetica Neue', Helvetica, Arial, sans-serif; font-weight: bold;">BUSINESS UNDERSTANDING</h2>
</div>



This project aims to improve marketing strategies, customize user experiences, and bolster security measures using Convolutional Neural Networks (CNNs). By analyzing facial images to predict gender, businesses can better understand their customer demographics, refine advertising efforts, and strengthen identity verification systems. Developing reliable CNN models enables data-driven decision-making, enhances operational efficiency, and fosters more meaningful interactions with customers.

<div style="background-color: #2D2D2D; color: #FFFFFF; padding: 10px; text-align: center; border-radius: 15px; box-shadow: 0 4px 8px rgba(0, 0, 0, 0);">
  <h2 style="margin: 0; font-family: 'Helvetica Neue', Helvetica, Arial, sans-serif; font-weight: bold;">Table of Contents</h2>
</div>


[1. Data Preprocessing](#-1-Data-Preprocessing)

[2. Training](#2-Training )

[3. Test Set Evaluation](#3-Test-Set-Evaluation)

[4. Summary](#3-Summary)

In [17]:
#import packages 

import os
import numpy as np
import matplotlib.pyplot as plt
from tensorflow.keras import layers
from tensorflow.keras import Model
from tensorflow.keras.preprocessing.image import ImageDataGenerator
import tensorflow as tf
from keras.optimizers import Adam
from keras.preprocessing import image

<div style="background-color: #2D2D2D; color: #FFFFFF; padding: 10px; text-align: center; border-radius: 15px; box-shadow: 0 4px 8px rgba(0, 0, 0, 0);">
  <h2 style="margin: 0; font-family: 'Helvetica Neue', Helvetica, Arial, sans-serif; font-weight: bold;">Data Preprocessing</h2>
</div>


In [18]:
# Data preprocessing for training images:
train_datagen = ImageDataGenerator(rescale = 1./255,
                                   rotation_range=40,
                                   width_shift_range=0.2,
                                   height_shift_range=0.2,
                                   shear_range=0.2,
                                   zoom_range=0.2,
                                   horizontal_flip=True,
                                   fill_mode='nearest')

# Data preprocessing for testing images:
test_datagen = ImageDataGenerator( rescale = 1.0/255)


# Training data generator:
train_generator = train_datagen.flow_from_directory("/Users/akshay/projects/Gender_classification/Dataset/Train",
                                                    batch_size =256 ,
                                                    class_mode = 'binary',
                                                    target_size = (64, 64))
# Validation data generator:
validation_generator =  test_datagen.flow_from_directory("/Users/akshay/projects/Gender_classification/Dataset/Validation",
                                                          batch_size  = 256,
                                                          class_mode  = 'binary',
                                                          target_size = (64, 64))

Found 160000 images belonging to 2 classes.
Found 22598 images belonging to 2 classes.


<div style="background-color: #2D2D2D; color: #FFFFFF; padding: 10px; text-align: center; border-radius: 15px; box-shadow: 0 4px 8px rgba(0, 0, 0, 0);">
  <h2 style="margin: 0; font-family: 'Helvetica Neue', Helvetica, Arial, sans-serif; font-weight: bold;">Model Training </h2>
</div>


In [12]:
# train and compile the neural network model for the task of Gender classification with Python:
model = tf.keras.models.Sequential([
    # 1st conv
    tf.keras.layers.Conv2D(96, (11,11),strides=(4,4), activation='relu', input_shape=(64, 64, 3)),
    tf.keras.layers.BatchNormalization(),
    tf.keras.layers.MaxPooling2D(2, strides=(2,2)),
    
    # 2nd conv
    tf.keras.layers.Conv2D(256, (11,11),strides=(1,1), activation='relu',padding="same"),
    tf.keras.layers.BatchNormalization(),
    
    # 3rd conv
    tf.keras.layers.Conv2D(384, (3,3),strides=(1,1), activation='relu',padding="same"),
    tf.keras.layers.BatchNormalization(),
    
    # 4th conv
    tf.keras.layers.Conv2D(384, (3,3),strides=(1,1), activation='relu',padding="same"),
    tf.keras.layers.BatchNormalization(),
    
    # 5th Conv
    tf.keras.layers.Conv2D(256, (3, 3), strides=(1, 1), activation='relu',padding="same"),
    tf.keras.layers.BatchNormalization(),
    tf.keras.layers.MaxPooling2D(2, strides=(2, 2)),
    
    # To Flatten layer
    tf.keras.layers.Flatten(),
    
    # To FC layer 1
    tf.keras.layers.Dense(4096, activation='relu'),
    tf.keras.layers.Dropout(0.5),
    
    #To FC layer 2
    tf.keras.layers.Dense(4096, activation='relu'),
    tf.keras.layers.Dropout(0.5),
    
    tf.keras.layers.Dense(1, activation='sigmoid')
])

# Compile the model with specified optimizer, loss function, and evaluation metrics
model.compile(
    optimizer=Adam(learning_rate=0.001), # Adam optimizer with learning rate 0.001
    loss='binary_crossentropy', # Binary cross-entropy loss function for binary classification task
    metrics=['accuracy'] # Accuracy metric to monitor during training
)

# Train the model using the specified training and validation generators
hist = model.fit(train_generator,
                 validation_data=validation_generator,
                 steps_per_epoch=256,
                 validation_steps=256,
                 epochs=50)

Epoch 1/50
256/256 [==============================] - ETA: 0s - loss: 0.9079 - accuracy: 0.5875WARNING:tensorflow:Your input ran out of data; interrupting training. Make sure that your dataset or generator can generate at least `steps_per_epoch * epochs` batches (in this case, 256 batches). You may need to use the repeat() function when building your dataset.


256/256 [==============================] - 911s 4s/step - loss: 0.9079 - accuracy: 0.5875 - val_loss: 0.5878 - val_accuracy: 0.6794
Epoch 2/50
256/256 [==============================] - 880s 3s/step - loss: 0.5482 - accuracy: 0.7235
Epoch 3/50
256/256 [==============================] - 955s 4s/step - loss: 0.4932 - accuracy: 0.7650
Epoch 4/50
256/256 [==============================] - 882s 3s/step - loss: 0.4515 - accuracy: 0.7917
Epoch 5/50
256/256 [==============================] - 839s 3s/step - loss: 0.4152 - accuracy: 0.8132
Epoch 6/50
256/256 [==============================] - 861s 3s/step - loss: 0.3719 - accuracy: 0.8352
Epoch 7/50
256/256 [==============================] - 1045s 4s/step - loss: 0.3422 - accuracy: 0.8504
Epoch 8/50
256/256 [==============================] - 1165s 5s/step - loss: 0.3153 - accuracy: 0.8646
Epoch 9/50
256/256 [==============================] - 1131s 4s/step - loss: 0.2928 - accuracy: 0.8732
Epoch 10/50
256/256 [==============================] - 97

In [29]:
# Save the trained model to a HDF5 file
model.save('gender_classification_model.h5')

<div style="background-color: #2D2D2D; color: #FFFFFF; padding: 10px; text-align: center; border-radius: 15px; box-shadow: 0 4px 8px rgba(0, 0, 0, 0);">
  <h2 style="margin: 0; font-family: 'Helvetica Neue', Helvetica, Arial, sans-serif; font-weight: bold;">Summary</h2>
</div>


The gender classification model was trained using a dataset consisting of 202,609 images. During the training process, the model achieved a loss of 0.1848 and an accuracy of 92.49%. These metrics indicate that the model has learned to effectively distinguish between male and female genders based on facial images. With a high accuracy rate and relatively low loss, the trained model demonstrates strong performance and reliability in gender prediction tasks.

The trained model has been saved as "**gender_classification_model.h5**"

To evaluate its performance, the model achieved 100% accuracy on the female dataset and 94% accuracy on the male dataset, resulting in a total accuracy of 97% on the test dataset.
